SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionLockTest.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "CrossServerPossessionLockTest.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "GameFramework/PlayerController.h"
+#include "Net/UnrealNetwork.h"
+#include "SpatialFunctionalTestFlowController.h"
+#include "TestPossessionPawn.h"
+#include "TestPossessionPlayerController.h"
+
+/**
+ * This test tests 1 locked Controller possess over 1 pawn.
+ *
+ * This test expects a load balancing grid and ACrossServerPossessionGameMode
+ * Recommand to use 2*2 load balancing grid because the position of Pawn was written in the code
+ * The client workers begin with a player controller and their default pawns, which they initially possess.
+ * The flow is as follows:
+ *  - Setup:
+ *    - Specify `GameMode Override` as ACrossServerPossessionGameMode
+ *    - Specify `Multi Worker Settings Class` as Zoning 2x2(e.g. BP_Possession_Settings_Zoning2_2 of UnrealGDKTestGyms)
+ *	  - Set `Num Required Clients` as 1
+ *  - Test:
+ *	  - Create a Pawn in first quadrant
+ *	  - Create Controller in other quadrant
+ *	  - Wait for Pawn in right worker.
+ *	  -	Lock the Controller and possess the Pawn
+ *	- Result Check:
+ *    - Pawn didn't have a Controller
+ */
+
+ACrossServerPossessionLockTest::ACrossServerPossessionLockTest()
+	: Super()
+{
+	Author = "Jay";
+	Description = TEXT("Test Locked Actor Cross-Server Possession");
+}
+
+void ACrossServerPossessionLockTest::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT("Controller remote possess"), FWorkerDefinition::AllClients, nullptr, nullptr, [this](float DeltaTime) {
+		ATestPossessionPawn* Pawn = GetPawn();
+		AssertIsValid(Pawn, TEXT("Test requires a Pawn"));
+		for (ASpatialFunctionalTestFlowController* FlowController : GetFlowControllers())
+		{
+			if (FlowController->WorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Client)
+			{
+				ATestPossessionPlayerController* Controller = Cast<ATestPossessionPlayerController>(FlowController->GetOwner());
+				if (Controller != nullptr)
+				{
+					Controller->RemotePossessOnClient(Pawn, true);
+				}
+			}
+		}
+		FinishStep();
+	});
+
+	AddWaitStep(FWorkerDefinition::AllServers);
+
+	AddStep(TEXT("Check test result"), FWorkerDefinition::AllServers, nullptr, nullptr, [this](float) {
```

Comment:
Worth running this on all workers?

Reply:
Yes, we can have only 1 server to check the results.
But we are not sure which server to check the results is best.
The Server index does not match the virtual worker id.